substrate/frame/assets-vesting/src/lib.rs

```diff
@@ -0,0 +1,786 @@
+			(frozen / duration).max(One::one())
+		};
+
+		let schedule = VestingInfo::new(frozen, per_block, starting_block);
+		debug_assert!(schedule.is_valid(), "merge_vesting_info schedule validation check failed");
+
+		Some(schedule)
+	}
+
+	// Execute a vested transfer from `source` to `target` with the given `schedule`.
+	fn do_vested_transfer(
+		asset: AssetIdOf<T, I>,
+		source: &T::AccountId,
+		target: &T::AccountId,
+		schedule: VestingInfo<BalanceOf<T, I>, BlockNumberFor<T>>,
+	) -> DispatchResult {
+		// Validate user inputs.
+		ensure!(schedule.locked() >= T::MinVestedTransfer::get(), Error::<T, I>::AmountLow);
+		if !schedule.is_valid() {
+			return Err(Error::<T, I>::InvalidScheduleParams.into())
+		};
+
+		// Check we can add to this account prior to any storage writes.
+		Self::can_add_vesting_schedule(
+			asset.clone(),
+			target,
+			schedule.locked(),
+			schedule.per_block(),
+			schedule.starting_block(),
+		)?;
+
+		T::Assets::transfer(
+			asset.clone(),
+			source,
+			target,
+			schedule.locked(),
+			Preservation::Expendable,
+		)?;
+
+		// We can't let this fail because the currency transfer has already happened.
+		// Must be successful as it has been checked before.
+		// Better to return error on failure anyway.
+		let res = Self::add_vesting_schedule(
+			asset,
+			target,
+			schedule.locked(),
+			schedule.per_block(),
+			schedule.starting_block(),
+		);
+		debug_assert!(res.is_ok(), "Failed to add a schedule when we had to succeed.");
+
+		Ok(())
+	}
+
+	/// Iterate through the schedules to track the current locked amount and
+	/// filter out completed and specified schedules.
+	///
+	/// Returns a tuple that consists of:
+	/// - Vec of vesting schedules, where completed schedules and those specified
+	/// 	by filter are removed. (Note the vec is not checked for respecting
+	/// 	bounded length.)
+	/// - The amount locked at the current block number based on the given schedules.
+	///
+	/// NOTE: the amount locked does not include any schedules that are filtered out via `action`.
+	fn report_schedule_updates(
+		schedules: Vec<VestingInfo<BalanceOf<T, I>, BlockNumberFor<T>>>,
+		action: VestingAction,
+	) -> (Vec<VestingInfo<BalanceOf<T, I>, BlockNumberFor<T>>>, BalanceOf<T, I>) {
+		let now = T::BlockNumberProvider::current_block_number();
+
+		let mut total_locked_now: BalanceOf<T, I> = Zero::zero();
+		let filtered_schedules = action
+			.pick_schedules::<T, I>(schedules)
+			.filter(|schedule| {
+				let locked_now = schedule.locked_at::<T::BlockNumberToBalance>(now);
+				let keep = !locked_now.is_zero();
+				if keep {
+					total_locked_now = total_locked_now.saturating_add(locked_now);
+				}
+				keep
+			})
+			.collect::<Vec<_>>();
+
+		(filtered_schedules, total_locked_now)
+	}
+
+	/// Write an accounts updated vesting lock to storage.
+	fn write_lock(
+		asset: AssetIdOf<T, I>,
+		who: &T::AccountId,
+		total_locked_now: BalanceOf<T, I>,
+	) -> DispatchResult {
+		T::Freezer::set_freeze(
```

Comment:
```suggestion
		T::Assets::set_freeze(
```
Because of the change I made above